### ISSUE TITLE
Darwin nebula client not finding/receiving udpAddrs for other hosts.

### ISSUE THREAD

**Issue by turtlewax** [slackhq/nebula#585]
My lighthouse most definitely knows and has a udpAddr for a remote host (Linux) and my mac host. Trouble is, my mac nebula host doesn't seem to be able to find out about the udpAddr for the other remote host (Linux).

Here is what I see on my lighthouse:
`Nov 15 10:26:14 hostname nebula[1478]: time="2021-11-15T10:26:14-05:00" level=info msg="Handshake message sent" certName=pc1 fingerprint=2733666f6007c2b508475b269fdb332a240343c49ba36b3973c5ffc31f999151 handshake="map[stage:2 style:ix_psk0]" initiatorIndex=1797597823 issuer=298e1986b9d3e1f20316a9b3389ebb037dcfe465932453e2b25c4f949e3e0217 remoteIndex=0 responderIndex=1842167964 sentCachedPackets=0 udpAddr="<redacted public IP>:48415" vpnIp=192.168.200.3`

And here is what my mac nebula client reports:
`INFO[0868] Handshake timed out durationNs=9382057076 handshake="map[stage:1 style:ix_psk0]" initiatorIndex=788921434 remoteIndex=0 udpAddrs="[]" vpnIp=192.168.200.3`

Now, the lighthouse log absolutely shows both of these separate hosts properly, with different, correct public udpAddrs assigned for each. The other remote host (192.168.200.3) is correctly reporting the udpAddr for the mac client, but the mac client never seems to get or find out about a udpAddr for the remote host (udpAddrs="[]", as seen above).

Why is this? Is this a bug in the Darwin nebula client that no one has reported yet?

**Comment by nbrownus** [comment]
Have you confirmed that the mac client is connected to the same lighthouse?

**Comment by turtlewax** [comment]
Yes, I only have a single lighthouse running.  The lighthouse itself has correct udpAddr and vpnIp values for both of the clients (Linux and Mac).  The Linux client does have a correct udpAddr value for the macOS client, but the macOS client has a perpetually empty udpAddr value for the Linux client.

Client configurations are basically identical, with the exception of certs and keys of course.

**Comment by nbrownus** [comment]
Do you have either of these setup on any of your hosts? https://github.com/slackhq/nebula/blob/78d0d46baea578c9d6e02c2f0f137068bc269839/examples/config.yml#L52
https://github.com/slackhq/nebula/blob/78d0d46baea578c9d6e02c2f0f137068bc269839/examples/config.yml#L76

**Comment by turtlewax** [comment]
No, those have been left commented out.  Additionally the nebula firewall is set as follows on both clients:

`  outbound:
    # Allow all outbound traffic from this node
    - port: any
      proto: any
      host: any

  inbound:
    # Allow all inbound from any nebula hosts
    - port: any
      proto: any
      host: any`

**Comment by nbrownus** [comment]
Any chance you can enable the `sshd` and run a `list-lighthouse-addrmap` on the lighthouse?

https://github.com/slackhq/nebula/blob/78d0d46baea578c9d6e02c2f0f137068bc269839/examples/config.yml#L132-L145

**Comment by turtlewax** [comment]
Yes - here is that command and it's redacted output:

```
root@nebula > list-lighthouse-addrmap
192.168.200.1: {"192.168.200.1":{"reported":[{"ip":"pub.ip.is.here","port":4242}]}}
192.168.200.2: {"192.168.200.2":{"learned":[{"ip":"pub.ip.is.here","port":54029}]}}
192.168.200.3: {"192.168.200.3":{"learned":[{"ip":"pub.ip.is.here","port":48415}],"reported":[{"ip":"192.168.0.113","port":48415}]}}
root@nebula >
```

The lighthouse has everything listed as I'd expect, public IPs, etc etc.

**Comment by nbrownus** [comment]
Interesting, this indicates that the mac host isn't reporting any ips to the lighthouse. Can you provide the config for that host as well as the first minute of logs after nebula starts?

**Comment by turtlewax** [comment]
macOS client config:

```
# This is the nebula example configuration file. You must edit, at a minimum, the static_host_map, lighthouse, and firewall sections
# Some options in this file are HUPable, including the pki section. (A HUP will reload credentials from disk without affecting existing tunnels)

# PKI defines the location of credentials for this node. Each of these can also be inlined by using the yaml ": |" syntax.
pki:
  # The CAs that are accepted by this node. Must contain one or more certificates created by 'nebula-cert ca'

#  ca: ca.crt
  ca: |
    -----BEGIN NEBULA CERTIFICATE-----
    SNIP
    -----END NEBULA CERTIFICATE-----

#  cert: macos.crt
  cert: |
    -----BEGIN NEBULA CERTIFICATE-----
    SNIP
    -----END NEBULA CERTIFICATE-----

#  key: macos.key
  key: |
    -----BEGIN NEBULA X25519 PRIVATE KEY-----
    SNIP
    -----END NEBULA X25519 PRIVATE KEY-----

  #blocklist is a list of certificate fingerprints that we will refuse to talk to
  #blocklist:
  #  - c99d4e650533b92061b09918e838a5a0a6aaee21eed1d12fd937682865936c72

# The static host map defines a set of hosts with fixed IP addresses on the internet (or any network).
# A host can have multiple fixed IP addresses defined here, and nebula will try each when establishing a tunnel.
# The syntax is:
#   "{nebula ip}": ["{routable ip/dns name}:{routable port}"]
# Example, if your lighthouse has the nebula IP of 192.168.100.1 and has the real ip address of 100.64.22.11 and runs on port 4242:
static_host_map:
  "192.168.200.1": ["light.house.pub.ip:4242"]


lighthouse:
  # am_lighthouse is used to enable lighthouse functionality for a node. This should ONLY be true on nodes
  # you have configured to be lighthouses in your network
  am_lighthouse: false
  # serve_dns optionally starts a dns listener that responds to various queries and can even be
  # delegated to for resolution
  #serve_dns: false
  #dns:
    # The DNS host defines the IP to bind the dns listener to. This also allows binding to the nebula node IP.
    #host: 0.0.0.0
    #port: 53
  # interval is the number of seconds between updates from this node to a lighthouse.
  # during updates, a node sends information about its current IP addresses to each node.
  interval: 60
  # hosts is a list of lighthouse hosts this node should report to and query from
  # IMPORTANT: THIS SHOULD BE EMPTY ON LIGHTHOUSE NODES
  # IMPORTANT2: THIS SHOULD BE LIGHTHOUSES' NEBULA IPs, NOT LIGHTHOUSES' REAL ROUTABLE IPs
hosts:
 - "192.168.200.1"

  # remote_allow_list allows you to control ip ranges that this node will
  # consider when handshaking to another node. By default, any remote IPs are
  # allowed. You can provide CIDRs here with `true` to allow and `false` to
  # deny. The most specific CIDR rule applies to each remote. If all rules are
  # "allow", the default will be "deny", and vice-versa. If both "allow" and
  # "deny" rules are present, then you MUST set a rule for "0.0.0.0/0" as the
  # default.
  #remote_allow_list:
    # Example to block IPs from this subnet from being used for remote IPs.
    #"172.16.0.0/12": false

    # A more complicated example, allow public IPs but only private IPs from a specific subnet
    #"0.0.0.0/0": true
    #"10.0.0.0/8": false
    #"10.42.42.0/24": true

  # local_allow_list allows you to filter which local IP addresses we advertise
  # to the lighthouses. This uses the same logic as `remote_allow_list`, but
  # additionally, you can specify an `interfaces` map of regular expressions
  # to match against interface names. The regexp must match the entire name.
  # All interface rules must be either true or false (and the default will be
  # the inverse). CIDR rules are matched after interface name rules.
  # Default is all local IP addresses.
  #local_allow_list:
    # Example to block tun0 and all docker interfaces.
    #interfaces:
      #tun0: false
      #'docker.*': false
    # Example to only advertise this subnet to the lighthouse.
    #"10.0.0.0/8": true

# Port Nebula will be listening on. The default here is 4242. For a lighthouse node, the port should be defined,
# however using port 0 will dynamically assign a port and is recommended for roaming nodes.
#listen:
#  host: 0.0.0.0
#  port: 4242
  # Sets the max number of packets to pull from the kernel for each syscall (under systems that support recvmmsg)
  # default is 64, does not support reload
  #batch: 64
  # Configure socket buffers for the udp side (outside), leave unset to use the system defaults. Values will be doubled by the kernel
  # Default is net.core.rmem_default and net.core.wmem_default (/proc/sys/net/core/rmem_default and /proc/sys/net/core/rmem_default)
  # Maximum is limited by memory in the system, SO_RCVBUFFORCE and SO_SNDBUFFORCE is used to avoid having to raise the system wide
  # max, net.core.rmem_max and net.core.wmem_max
  #read_buffer: 10485760
  #write_buffer: 10485760

punchy:
  # Continues to punch inbound/outbound at a regular interval to avoid expiration of firewall nat mappings
  punch: true

  # respond means that a node you are trying to reach will connect back out to you if your hole punching fails
  # this is extremely useful if one node is behind a difficult nat, such as a symmetric NAT
  # Default is false
  respond: true

  # delays a punch response for misbehaving NATs, default is 1 second, respond must be true to take effect
  delay: 1s

# Cipher allows you to choose between the available ciphers for your network.
# IMPORTANT: this value must be identical on ALL NODES/LIGHTHOUSES. We do not/will not support use of different ciphers simultaneously!
#cipher: chachapoly

# Local range is used to define a hint about the local network range, which speeds up discovering the fastest
# path to a network adjacent nebula node.
#local_range: "172.16.0.0/24"

# sshd can expose informational and administrative functions via ssh this is a
#sshd:
  # Toggles the feature
  #enabled: true
  # Host and port to listen on, port 22 is not allowed for your safety
  #listen: 127.0.0.1:2222
  # A file containing the ssh host private key to use
  # A decent way to generate one: ssh-keygen -t ed25519 -f ssh_host_ed25519_key -N "" < /dev/null
  #host_key: ./ssh_host_ed25519_key
  # A file containing a list of authorized public keys
  #authorized_users:
    #- user: steeeeve
      # keys can be an array of strings or single string
      #keys:
        #- "ssh public key string"

# Configure the private interface. Note: addr is baked into the nebula certificate
tun:
  # When tun is disabled, a lighthouse can be started without a local tun interface (and therefore without root)
  disabled: false
  # Name of the device
  dev: utun10
  # Toggles forwarding of local broadcast packets, the address of which depends on the ip/mask encoded in pki.cert
  drop_local_broadcast: true
  # Toggles forwarding of multicast packets
  drop_multicast: false
  # Sets the transmit queue length, if you notice lots of transmit drops on the tun it may help to raise this number. Default is 500
  tx_queue: 500
  # Default MTU for every packet, safe setting is (and the default) 1300 for internet based traffic
  mtu: 1300
  # Route based MTU overrides, you have known vpn ip paths that can support larger MTUs you can increase/decrease them here
  routes:
    #- mtu: 8800
    #  route: 10.0.0.0/16
  # Unsafe routes allows you to route traffic over nebula to non-nebula nodes
  # Unsafe routes should be avoided unless you have hosts/services that cannot run nebula
  # NOTE: The nebula certificate of the "via" node *MUST* have the "route" defined as a subnet in its certificate
  unsafe_routes:
    #- route: 172.16.1.0/24
    #  via: 192.168.100.99
    #  mtu: 1300 #mtu will default to tun mtu if this option is not sepcified


# TODO
# Configure logging level
logging:
  # panic, fatal, error, warning, info, or debug. Default is info
  level: info
  # json or text formats currently available. Default is text
  format: text
  # Disable timestamp logging. useful when output is redirected to logging system that already adds timestamps. Default is false
  #disable_timestamp: true
  # timestamp format is specified in Go time format, see:
  #     https://golang.org/pkg/time/#pkg-constants
  # default when `format: json`: "2006-01-02T15:04:05Z07:00" (RFC3339)
  # default when `format: text`:
  #     when TTY attached: seconds since beginning of execution
  #     otherwise: "2006-01-02T15:04:05Z07:00" (RFC3339)
  # As an example, to log as RFC3339 with millisecond precision, set to:
  #timestamp_format: "2006-01-02T15:04:05.000Z07:00"

#stats:
  #type: graphite
  #prefix: nebula
  #protocol: tcp
  #host: 127.0.0.1:9999
  #interval: 10s

  #type: prometheus
  #listen: 127.0.0.1:8080
  #path: /metrics
  #namespace: prometheusns
  #subsystem: nebula
  #interval: 10s

  # enables counter metrics for meta packets
  #   e.g.: `messages.tx.handshake`
  # NOTE: `message.{tx,rx}.recv_error` is always emitted
  #message_metrics: false

  # enables detailed counter metrics for lighthouse packets
  #   e.g.: `lighthouse.rx.HostQuery`
  #lighthouse_metrics: false

# Handshake Manger Settings
#handshakes:
  # Total time to try a handshake = sequence of `try_interval * retries`
  # With 100ms interval and 20 retries it is 23.5 seconds
  #try_interval: 100ms
  #retries: 20
  # wait_rotation is the number of handshake attempts to do before starting to try non-local IP addresses
  #wait_rotation: 5
  # trigger_buffer is the size of the buffer channel for quickly sending handshakes
  # after receiving the response for lighthouse queries
  #trigger_buffer: 64

# Nebula security group configuration
firewall:
  conntrack:
    tcp_timeout: 12m
    udp_timeout: 3m
    default_timeout: 10m
    max_connections: 100000

  # The firewall is default deny. There is no way to write a deny rule.
  # Rules are comprised of a protocol, port, and one or more of host, group, or CIDR
  # Logical evaluation is roughly: port AND proto AND (ca_sha OR ca_name) AND (host OR group OR groups OR cidr)
  # - port: Takes `0` or `any` as any, a single number `80`, a range `200-901`, or `fragment` to match second and further fragments of fragmented packets (since there is no port available).
  #   code: same as port but makes more sense when talking about ICMP, TODO: this is not currently implemented in a way that works, use `any`
  #   proto: `any`, `tcp`, `udp`, or `icmp`
  #   host: `any` or a literal hostname, ie `test-host`
  #   group: `any` or a literal group name, ie `default-group`
  #   groups: Same as group but accepts a list of values. Multiple values are AND'd together and a certificate would have to contain all groups to pass
  #   cidr: a CIDR, `0.0.0.0/0` is any.
  #   ca_name: An issuing CA name
  #   ca_sha: An issuing CA shasum

  outbound:
    # Allow all outbound traffic from this node
    - port: any
      proto: any
      host: any

  inbound:
    # Allow all inbound from any nebula hosts
    - port: any
      proto: any
      host: any

    # Allow tcp/443 from any host with BOTH laptop and home group
    # - port: 443
    #   proto: tcp
    #   groups:
    #     - laptop
    #     - home
```

**Comment by turtlewax** [comment]
First minute of output from macOS client.  *note* I manually pinged the lighthouse (192.168.200.1) and attempted to ping the Linux client so you could observe results:

```
INFO[0000] Firewall rule added                           firewallRule="map[caName: caSha: direction:outgoing endPort:0 groups:[] host:any ip: proto:0 startPort:0]"
INFO[0000] Firewall rule added                           firewallRule="map[caName: caSha: direction:incoming endPort:0 groups:[] host:any ip: proto:0 startPort:0]"
INFO[0000] Firewall started                              firewallHash=21716b47a7a140e448077fe66c31b4b42f232e996818d7dd1c6c4991e066dbdb
INFO[0000] Main HostMap created                          network=192.168.200.2/24 preferredRanges="[]"
INFO[0000] UDP hole punching enabled
INFO[0000] Nebula interface is active                    build=1.5.0 interface=utun10 network=192.168.200.2/24 udpAddr="[::]:59433"
INFO[0016] Handshake message sent                        handshake="map[stage:1 style:ix_psk0]" initiatorIndex=105360011 udpAddrs="[light.house.pub.ip:4242]" vpnIp=192.168.200.1
INFO[0016] Handshake message received                    certName=lighthouse1 durationNs=17793308 fingerprint=80a9208a72129a5ceaaa33e9382bb98b90ef8d51cc1d7ae8d9fbac494e3843c1 handshake="map[stage:2 style:ix_psk0]" initiatorIndex=105360011 issuer=298e1986b9d3e1f20316a9b3389ebb037dcfe465932453e2b25c4f949e3e0217 remoteIndex=105360011 responderIndex=3730905281 sentCachedPackets=1 udpAddr="light.house.pub.ip:4242" vpnIp=192.168.200.1
INFO[0038] Handshake timed out                           durationNs=9407001582 handshake="map[stage:1 style:ix_psk0]" initiatorIndex=1100610786 remoteIndex=0 udpAddrs="[]" vpnIp=192.168.200.3
```

**Comment by nbrownus** [comment]
Looks like a yaml issue. `hosts` should be indented under `lighthouse`.

```
lighthouse:
  # am_lighthouse is used to enable lighthouse functionality for a node. This should ONLY be true on nodes
  # you have configured to be lighthouses in your network
  am_lighthouse: false
  # serve_dns optionally starts a dns listener that responds to various queries and can even be
  # delegated to for resolution
  #serve_dns: false
  #dns:
    # The DNS host defines the IP to bind the dns listener to. This also allows binding to the nebula node IP.
    #host: 0.0.0.0
    #port: 53
  # interval is the number of seconds between updates from this node to a lighthouse.
  # during updates, a node sends information about its current IP addresses to each node.
  interval: 60
  # hosts is a list of lighthouse hosts this node should report to and query from
  # IMPORTANT: THIS SHOULD BE EMPTY ON LIGHTHOUSE NODES
  # IMPORTANT2: THIS SHOULD BE LIGHTHOUSES' NEBULA IPs, NOT LIGHTHOUSES' REAL ROUTABLE IPs
  hosts:
   - "192.168.200.1"
  ```

**Comment by turtlewax** [comment]
@nbrownus Gah! (looks around sheepishly).... Sorry for the false alarm.  This indeed *was* the issue; I've indented the `hosts:` entry under `lighthouse` as you specified, and it works now!  I figured Nebula would fail to start or throw an error if I had a floating `hosts:` entry in the .yaml, but I guess a top level entry for that is allowed?

If nothing else I've learned a few more tricks about troubleshooting Nebula!

Apologies again for the wasted time, Nate.
